[clang-include-cleaner] Fix incorrect directory issue for writing files

bc-lee opened the pull request:
If the current working directory of `clang-include-cleaner` differs from the directory of the input files specified in the compilation database, it doesn't adjust the input file paths properly. As a result, `clang-include-cleaner` either writes files to the wrong directory or fails to write files altogether.

This pull request fixes the issue by adjusting the input file paths based on the directory specified in the compilation database. If that directory is not writable, `clang-include-cleaner` will write the output relative to the current working directory.

Fixes #110843.

bc-lee committed:
[clang-include-cleaner] Fix incorrect directory issue for writing files

If the current working directory of `clang-include-cleaner` differs from
the directory of the input files specified in the compilation database,
it doesn't adjust the input file paths properly. As a result,
`clang-include-cleaner` either writes files to the wrong directory or
fails to write files altogether.

This pull request fixes the issue by checking whether the input file
path is absolute. If it isn't, the path is concatenated with
the directory of the input file specified in the compilation database.

Fixes #110843.

bc-lee commented:
@kadircet @hokein Could you please review this PR? It is related to #102615. More details are provided in the related issue. Thanks!

kadircet reviewed (changes requested):
thanks a lot for the patch, and sorry for regressing this.

I think this is also going to regress behavior in some cases (e.g. working directory in compile commands might be virtual for some build systems like bazel).

When include-cleaner is invoked as `clang-include-cleaner foo.cc` the user is definitely trying to run this on a file relative to directory in which we invoked include-cleaner, hence we should make sure it's adressed through that path going forward.

Can we instead update logic near the following in `main`:
```cpp
 clang::tooling::ClangTool Tool(OptionsParser->getCompilations(),
                                 OptionsParser->getSourcePathList());
```

instead of passing `OptionsParser->getSourcePathList())`, we can turn all of these source paths into absolue paths, using the current-working-dir of the process, and then we don't need to worry about how they make their way into our ast-consumer.

bc-lee commented:
Could you provide a minimal reproducer for this case if possible? Ideally, it should not depend on Google's internal Bazel setup or tools. From what I understand, Bazel itself doesn’t generate compilation databases, and third-party tools adjust the paths of raw compiler invocations to the correct ones. I tested two methods for generating compilation databases for Bazel projects in my demo repository (https://github.com/bc-lee/test-compilation-database), and both methods correctly generated the compilation databases, with no issues traversing over the Bazel build directory.

For instance, when I ran [kiron1/bazel-compile-commands](https://github.com/kiron1/bazel-compile-commands) in my demo repository, it generated the following `compile_commands.json`:

<details>
<summary>Click to toggle</summary>

```json
[
  {
    "directory": "/home/leebc/.cache/bazel/_bazel_leebc/5ea2deb937e75f0bf29ba4c49931e67d/execroot/_main",
    "command": "/usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -std=c++14 -MD -MF bazel-out/k8-fastbuild/bin/api/_objs/main/main.pic.d -frandom-seed=bazel-out/k8-fastbuild/bin/api/_objs/main/main.pic.o -fPIC -iquote . -iquote bazel-out/k8-fastbuild/bin -iquote external/bazel_tools -iquote bazel-out/k8-fastbuild/bin/external/bazel_tools -std=c++17 -fno-canonical-system-headers -Wno-builtin-macro-redefined -D__DATE__=\"redacted\" -D__TIMESTAMP__=\"redacted\" -D__TIME__=\"redacted\" -c api/main.cc -o bazel-out/k8-fastbuild/bin/api/_objs/main/main.pic.o",
    "file": "api/main.cc",
    "output": "bazel-out/k8-fastbuild/bin/api/_objs/main/main.pic.o"
  }
]
```

</details>

And `clang-include-cleaner` ran correctly. The file path was `api/main.cc`, and the compiler invocation’s working directory was `/home/leebc/.cache/bazel/_bazel_leebc/5ea2deb937e75f0bf29ba4c49931e67d/execroot/_main`, so `clang-include-cleaner` correctly adjusted the path to `/home/leebc/.cache/bazel/_bazel_leebc/5ea2deb937e75f0bf29ba4c49931e67d/execroot/_main/api/main.cc`, which was correctly symlinked.

If there are any other cases that I missed, please let me know. I’ll try to reproduce the issue and fix it.

kadircet commented:
> so clang-include-cleaner correctly adjusted the path to /home/leebc/.cache/bazel/_bazel_leebc/5ea2deb937e75f0bf29ba4c49931e67d/execroot/_main/api/main.cc, which was correctly symlinked.

in our case this is symlinked to an immutable copy of the source file (to make sure builds don't act weird if you edit some sources while the build is running). hence it'll be a regression.

What's the concern about not performing this directly as relative to the process' working directory? That approach doesn't rely on any assumptions about the build system, or filesystem layout hence feels much more robust.

bc-lee committed:
Fix path issue identified by @kadircet

Some Bazel systems might have a read-only file system in the build
directory. In such cases, the output file should be written to
the original source file. Adjusted the logic to find the correct
output directory based on the compilation database and
the current working directory.

bc-lee commented:
> in our case this is symlinked to an immutable copy of the source file (to make sure builds don't act weird if you edit some sources while the build is running). hence it'll be a regression.
> 
> What's the concern about not performing this directly as relative to the process' working directory? That approach doesn't rely on any assumptions about the build system, or filesystem layout hence feels much more robust.

It’s unfortunate that we can't apply unified logic for determining the actual output file. It's especially disappointing given that I’m using GN and you’re using Bazel, both of which are build systems from Google.

I’ve adjusted the logic to find the correct output directory before the compiler invocation starts. Please take another look and let me know if there's anything else I should consider.

kadircet commented on clang-tools-extra/include-cleaner/tool/IncludeCleaner.cpp:
```diff
@@ -305,7 +342,32 @@ int main(int argc, const char **argv) {
   }
 
-  clang::tooling::ClangTool Tool(OptionsParser->getCompilations(),
+  auto &CompilationDatabase = OptionsParser->getCompilations();
```
i think this can be simplified with something like:
```
auto VFS = llvm::vfs::getRealFileSystem();
auto &CDB = OptionsParser->getCompilations();
std::map<std::string, std::string> CDBToAbsPaths;
for (auto &Source : OptionsParser->getSourcePathList()) {
  llvm::SmallString<256> AbsPath(Source);
  if (auto Err = VFS.makeAbsolute(AbsPath)) {
      llvm::errs() << "Failed to get absolute path for " << Source << " : " <<  Err.message() << '\n';
      return 1;
  }
  for(auto Cmd : CDB.getCompilecommands(AbsPath)) {
    llvm::SmallString<256> CDBPath(Cmd.Filename);
    llvm::sys::fs::make_absolute(Cmd.Directory, CDBPath);
    CDBToAbsPaths[CDBPath] = AbsPath;
  }
}
....
if (Edit) {
  for (const auto &NameAndContent : Factory.editedFiles()) {
     llvm::StringRef FileName = NameAndContent.first();
     if (auto It = CDBToAbsPaths.find(FileName); It != CDBToAbsPaths.end())
         FileName = It->second;
     ...
  }
}
```

bc-lee replied:
I didn't simplify it much, but I tried to explain my intent as clearly as possible. However, I feel the code is somewhat verbose.

kadircet commented on clang-tools-extra/include-cleaner/tool/IncludeCleaner.cpp:
```diff
@@ -173,6 +173,7 @@ class Action : public clang::ASTFrontendAction {
       writeHTML();
 
+    // Source File's path relative to compilation database.
     llvm::StringRef Path =
```
can we keep reporting with abs-paths here, i.e:
```
llvm::SmallString<256> AbsPath(SM.getFileEntryRefForID(SM.getMainFileID())->getName());
SM.getFileManager().makeAbsolutePath(AbsPath);
...
```

This way we can recognize these paths in main and perform the mapping accordingly.

kadircet reviewed (commented):
thanks a lot for the revision!

One concern is, we can't really use `getAllCompileCommands()` reliably either, not all compilation databases implement that, and even if they do this is gonna be an overkill. For example for chromium we'll iterate over tens of thousands of compile flags, just to modify a handful of files. I've made some suggestions in line.

Also I think it's easier if we put absolute-paths from CDB into the `EditedFiles` and just remap them to absolute paths per tools working directory afterwards when we can.

bc-lee committed:
Apply code review feedback

- Store the compilation database file as absolute path
- Instead of using `llvm::sys::fs::current_path`,
  use `FileSystem::makeAbsolute` to get the absolute path
  of various files
- Add more comments to explain the intent of the code

bc-lee committed:
Add test cases where the cwd and the compilation database's directory mismatch

bc-lee committed:
Fix the previous commit that removed existing test cases

bc-lee committed:
Make clang-format happy

bc-lee committed:
Fix unittest failures in Windows

kadircet commented on clang-tools-extra/include-cleaner/tool/IncludeCleaner.cpp:
```diff
@@ -305,8 +307,84 @@ int main(int argc, const char **argv) {
+      llvm::SmallString<256> CDBPath(Cmd.Filename);
+      std::string Directory(Cmd.Directory);
+
+      if (llvm::sys::path::is_absolute(CDBPath)) {
+        // If the path in the compilation database is already absolute, we don't
+        // need to do anything.
+        CDBToAbsPaths[static_cast<std::string>(CDBPath)] =
+            static_cast<std::string>(AbsPath);
+      } else {
+        auto Sept = llvm::sys::path::get_separator();
+        // First, try to find the file based on the compilation database.
+        llvm::Twine FilePathTwine = Directory + Sept + CDBPath;
+        llvm::SmallString<256> FilePath;
+        FilePathTwine.toVector(FilePath);
+        // Check if it is writable.
+        if (llvm::sys::fs::access(FilePath, llvm::sys::fs::AccessMode::Write) !=
+            std::error_code()) {
+          // If not, try to find the file based on the current working
+          // directory, as some Bazel invocations may not set the compilation
+          // invocation's filesystem as non-writable. In such cases, we can
+          // find the file based on the current working directory.
+          FilePath = Source;
+          if (auto EC = VFS->makeAbsolute(CDBPath)) {
+            llvm::errs() << "Failed to get absolute path for " << CDBPath
+                         << " : " << EC.message() << '\n';
+            return 1;
+          }
+          if (llvm::sys::fs::access(FilePath,
+                                    llvm::sys::fs::AccessMode::Write) !=
+              std::error_code()) {
+            llvm::errs() << "Failed to find a writable path for " << Source
+                         << '\n';
+            return 1;
+          }
+        }
+        CDBToAbsPaths[static_cast<std::string>(CDBPath)] =
+            static_cast<std::string>(FilePath);
+      }
+    }
+  }
```
`llvm::sys::fs::make_absolute(Cmd.Directory, CDBPath);` already performs concatenation only if `CDBPath` is not an absolute path, you don't re-implement all of this.

bc-lee replied:
Done.

kadircet commented on clang-tools-extra/include-cleaner/tool/IncludeCleaner.cpp:
```diff
@@ -305,8 +307,84 @@ int main(int argc, const char **argv) {
+      if (llvm::sys::path::is_absolute(CDBPath)) {
+        // If the path in the compilation database is already absolute, we don't
+        // need to do anything.
+        CDBToAbsPaths[static_cast<std::string>(CDBPath)] =
```
we prefer `std::string(CDBPath)` or `CDBPath.str().str()` to `static_cast`. (also for other occurences of `static_cast`)

bc-lee replied:
Done.

kadircet commented on clang-tools-extra/include-cleaner/tool/IncludeCleaner.cpp:
```diff
@@ -305,8 +307,84 @@ int main(int argc, const char **argv) {
+        llvm::SmallString<256> FilePath;
+        FilePathTwine.toVector(FilePath);
+        // Check if it is writable.
+        if (llvm::sys::fs::access(FilePath, llvm::sys::fs::AccessMode::Write) !=
+            std::error_code()) {
```
i don't see why we need to complicate the logic by checking for this. is there any reasons to not always map to file-path relative to current process?

i am pretty convinced that the user intention is to almost always edit the file path as referred to during invocation, e.g. `cd /foo && clang-include-cleaner path/to/foo.cc` is always meant to edit `/foo/path/to/foo.cc` no matter what file path we get from CDB.

i can see how this might also work, but I prefer to maintain code that has as few special cases as possible. so unless something is indeed breaking with this simplification, can you please get rid of this case?

bc-lee replied:
It turns out we don't need to check for this. I removed those checks.

kadircet commented on clang-tools-extra/include-cleaner/test/tool.cpp:
```diff
@@ -48,3 +48,13 @@ int x = foo();
 //        RUN: FileCheck --match-full-lines --check-prefix=EDIT2 %s < %t.cpp
 //  EDIT2-NOT: {{^}}#include "foo.h"{{$}}
+
+//        RUN: mkdir -p $(dirname %t)/out
```
you can use `%T` instead of `dirname %t`, you can find full list in https://llvm.org/docs/CommandGuide/lit.html#substitutions, same below.

also it's safer to run a `rm -rf %T` beforehand to make sure we're starting clean

bc-lee replied:
It turns out `%T` is also deprecated. I replaced my shell invocation with `%t.dir`, as this is the recommended method, as described in [https://reviews.llvm.org/D69572](https://reviews.llvm.org/D69572). I also slightly modified the test invocation so that it works on Windows as well. (Tested on a local Windows machine.)

kadircet commented on clang-tools-extra/include-cleaner/test/tool.cpp:
```diff
@@ -49,12 +49,15 @@ int x = foo();
 
+// This test has commands that rely on shell capabilities that won't execute
+// correctly on Windows e.g. subshell execution
+//   REQUIRES: shell
```
this doesn't just affect that particular run, affects the whole, can you move the test into its own file ?

bc-lee replied:
This was explained in the previous comment. It now works on Windows as well.

bc-lee committed:
Apply review comments.

- Removed the check for Edit mode.
- Removed unnecessary fallback.
- Various cleanups.
- Make the test work on Windows as well.

bc-lee commented on clang-tools-extra/include-cleaner/tool/IncludeCleaner.cpp:
```diff
@@ -305,8 +307,84 @@ int main(int argc, const char **argv) {
+  // CDBToAbsPaths is a map from the path in the compilation database to the
+  // writable absolute path of the file.
+  std::map<std::string, std::string> CDBToAbsPaths;
+  if (Edit) {
```
Done. Removed the check for Edit mode.

bc-lee commented on clang-tools-extra/include-cleaner/tool/IncludeCleaner.cpp:
```diff
@@ -305,8 +307,84 @@ int main(int argc, const char **argv) {
+        if (Cmds.empty()) {
+          continue;
+        }
+      }
```
Done. I checked, and even if we don't specify a compilation database, clang will create one using the `FixedCompilationDatabase` class. So, I added an error and return early if the compilation database can't be found.

bc-lee commented on clang-tools-extra/include-cleaner/tool/IncludeCleaner.cpp:
```diff
@@ -305,8 +307,84 @@ int main(int argc, const char **argv) {
+        }
+      }
+      // We only need the first command to get the directory.
+      auto Cmd = Cmds[0];
```
Done.

kadircet commented on clang-tools-extra/include-cleaner/tool/IncludeCleaner.cpp:
```diff
@@ -305,8 +307,46 @@ int main(int argc, const char **argv) {
+  // CDBToAbsPaths is a map from the path in the compilation database to the
+  // writable absolute path of the file.
+  std::map<std::string, std::string> CDBToAbsPaths;
+  // if Edit is enabled, `Factory.editedFiles()` will contain the final code,
```
can you also drop the mention of `Edit` from the comment?

kadircet commented on clang-tools-extra/include-cleaner/tool/IncludeCleaner.cpp:
```diff
@@ -305,8 +307,46 @@ int main(int argc, const char **argv) {
+    std::vector<clang::tooling::CompileCommand> Cmds =
+        CDB.getCompileCommands(AbsPath);
+    if (Cmds.empty()) {
+      // Try with the original path.
+      Cmds = CDB.getCompileCommands(Source);
```
this fallback is still not needed. if we can't find the flags with absolute path, it's fine to bail out.

kadircet commented on clang-tools-extra/include-cleaner/tool/IncludeCleaner.cpp:
```diff
@@ -305,8 +307,46 @@ int main(int argc, const char **argv) {
+  // absolute or relative, and if it is relative, it is relative to the
+  // "Directory" field in the compilation database. We need to make it
+  // absolute to write the final code to the correct path.
+  for (auto &Source : OptionsParser->getSourcePathList()) {
+    llvm::SmallString<256> AbsPath(Source);
+    if (auto Err = VFS->makeAbsolute(AbsPath)) {
+      llvm::errs() << "Failed to get absolute path for " << Source << " : "
+                   << Err.message() << '\n';
+      return 1;
+    }
+    std::vector<clang::tooling::CompileCommand> Cmds =
+        CDB.getCompileCommands(AbsPath);
+    if (Cmds.empty()) {
+      // Try with the original path.
+      Cmds = CDB.getCompileCommands(Source);
+      if (Cmds.empty()) {
+        // It should be found in the compilation database, even user didn't
+        // specify the compilation database, the `FixedCompilationDatabase` will
+        // create an entry from the arguments. So it is an error if we can't
+        // find the compile commands.
+        llvm::errs() << "No compile commands found for " << Source << '\n';
+        return 1;
+      }
+    }
+    for (const auto &Cmd : Cmds) {
+      llvm::SmallString<256> CDBPath(Cmd.Filename);
+      std::string Directory(Cmd.Directory);
+      llvm::sys::fs::make_absolute(Cmd.Directory, CDBPath);
+      CDBToAbsPaths[std::string(CDBPath)] = std::string(AbsPath);
+    }
+  }
```
nit: you can extract all of this logic to a helper function:
```cpp
llvm::Expected<std::map<...>> mapInputsToAbsPaths(...);
...
int main(...) {
  ...
  auto CDBToAbsPaths = mapInputsToAbsPaths(CDB, Opts->OptionsParser->getSourcePathList());
  if (!CDBToAbsPaths) {
    return 1;
  }
  ...
}

kadircet reviewed (approved):
thanks for bearing with me, LGTM!

bc-lee committed:
Apply review comments

- Removed unnecessary mention of Edit
- Removed fallback logic when the absolute path isn't found
- Extracted mapInputsToAbsPaths as a separate function

bc-lee commented:
Thanks for the review!

kadircet commented:
btw, LMK if you don't have commit access and I should merge this for you

bc-lee commented:
I don't have commit access to LLVM. Please merge this PR. Thanks!

bc-lee commented:
I will fix the following error within an hour (currently away from keyboard).

```
FAILED: tools/clang/tools/extra/include-cleaner/tool/CMakeFiles/clang-include-cleaner.dir/IncludeCleaner.cpp.o 
ccache /home/docker/llvm-external-buildbots/clang.17.0.6/bin/clang++ --gcc-toolchain=/gcc-toolchain/usr -DGTEST_HAS_RTTI=0 -DLLVM_BUILD_STATIC -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/build/tools/clang/tools/extra/include-cleaner/tool -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/clang-tools-extra/include-cleaner/tool -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/clang/include -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/build/tools/clang/include -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/build/include -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/llvm/include -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/clang-tools-extra/include-cleaner/include -I/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/clang-tools-extra/include-cleaner/tool/../lib -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -fno-common -Woverloaded-virtual -Wno-nested-anon-types -O3 -DNDEBUG -std=c++17  -fno-exceptions -funwind-tables -fno-rtti -UNDEBUG -MD -MT tools/clang/tools/extra/include-cleaner/tool/CMakeFiles/clang-include-cleaner.dir/IncludeCleaner.cpp.o -MF tools/clang/tools/extra/include-cleaner/tool/CMakeFiles/clang-include-cleaner.dir/IncludeCleaner.cpp.o.d -o tools/clang/tools/extra/include-cleaner/tool/CMakeFiles/clang-include-cleaner.dir/IncludeCleaner.cpp.o -c /home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/clang-tools-extra/include-cleaner/tool/IncludeCleaner.cpp
/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/clang-tools-extra/include-cleaner/tool/IncludeCleaner.cpp:302:14: error: moving a temporary object prevents copy elision [-Werror,-Wpessimizing-move]
  302 |       return std::move(llvm::errorCodeToError(Err));
      |              ^
/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/clang-tools-extra/include-cleaner/tool/IncludeCleaner.cpp:302:14: note: remove std::move call here
  302 |       return std::move(llvm::errorCodeToError(Err));
      |              ^~~~~~~~~~                           ~
1 error generated.
```
